File: tests/rosetta/transpiler/FS/b-zier-curves-intersections.fs
// Generated 2025-07-26 05:05 +0700

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
type Point = {
    x: float
    y: float
}
type QuadSpline = {
    c0: float
    c1: float
    c2: float
}
type QuadCurve = {
    x: QuadSpline
    y: QuadSpline
}
type Anon1 = {
    p: obj
    q: obj
}
let rec absf (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        __ret <- if x < 0.0 then (-x) else x
        raise Return
        __ret
    with
        | Return -> __ret
and maxf (a: float) (b: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable a = a
    let mutable b = b
    try
        __ret <- if a > b then a else b
        raise Return
        __ret
    with
        | Return -> __ret
and minf (a: float) (b: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable a = a
    let mutable b = b
    try
        __ret <- if a < b then a else b
        raise Return
        __ret
    with
        | Return -> __ret
and max3 (a: float) (b: float) (c: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable a = a
    let mutable b = b
    let mutable c = c
    try
        let mutable m: float = a
        if b > m then
            m <- b
        if c > m then
            m <- c
        __ret <- m
        raise Return
        __ret
    with
        | Return -> __ret
and min3 (a: float) (b: float) (c: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable a = a
    let mutable b = b
    let mutable c = c
    try
        let mutable m: float = a
        if b < m then
            m <- b
        if c < m then
            m <- c
        __ret <- m
        raise Return
        __ret
    with
        | Return -> __ret
and subdivideQuadSpline (q: QuadSpline) (t: float) =
    let mutable __ret : QuadSpline array = Unchecked.defaultof<QuadSpline array>
    let mutable q = q
    let mutable t = t
    try
        let s: float = 1.0 - t
        let mutable u: QuadSpline = { c0 = q.c0; c1 = 0.0; c2 = 0.0 }
        let mutable v: QuadSpline = { c0 = 0.0; c1 = 0.0; c2 = q.c2 }
        u <- { u with c1 = (s * (q.c0)) + (t * (q.c1)) }
        v <- { v with c1 = (s * (q.c1)) + (t * (q.c2)) }
        u <- { u with c2 = (s * (u.c1)) + (t * (v.c1)) }
        v <- { v with c0 = u.c2 }
        __ret <- [|u; v|]
        raise Return
        __ret
    with
        | Return -> __ret
and subdivideQuadCurve (q: QuadCurve) (t: float) =
    let mutable __ret : QuadCurve array = Unchecked.defaultof<QuadCurve array>
    let mutable q = q
    let mutable t = t
    try
        let xs: QuadSpline array = subdivideQuadSpline (q.x) t
        let ys: QuadSpline array = subdivideQuadSpline (q.y) t
        let mutable u: QuadCurve = { x = xs.[0]; y = ys.[0] }
        let mutable v: QuadCurve = { x = xs.[1]; y = ys.[1] }
        __ret <- [|u; v|]
        raise Return
        __ret
    with
        | Return -> __ret
and rectsOverlap (xa0: float) (ya0: float) (xa1: float) (ya1: float) (xb0: float) (yb0: float) (xb1: float) (yb1: float) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable xa0 = xa0
    let mutable ya0 = ya0
    let mutable xa1 = xa1
    let mutable ya1 = ya1
    let mutable xb0 = xb0
    let mutable yb0 = yb0
    let mutable xb1 = xb1
    let mutable yb1 = yb1
    try
        __ret <- (((xb0 <= xa1) && (xa0 <= xb1)) && (yb0 <= ya1)) && (ya0 <= yb1)
        raise Return
        __ret
    with
        | Return -> __ret
and testIntersect (p: QuadCurve) (q: QuadCurve) (tol: float) =
    let mutable __ret : Map<string, obj> = Unchecked.defaultof<Map<string, obj>>
    let mutable p = p
    let mutable q = q
    let mutable tol = tol
    try
        let pxmin: float = min3 (p.x.c0) (p.x.c1) (p.x.c2)
        let pymin: float = min3 (p.y.c0) (p.y.c1) (p.y.c2)
        let pxmax: float = max3 (p.x.c0) (p.x.c1) (p.x.c2)
        let pymax: float = max3 (p.y.c0) (p.y.c1) (p.y.c2)
        let qxmin: float = min3 (q.x.c0) (q.x.c1) (q.x.c2)
        let qymin: float = min3 (q.y.c0) (q.y.c1) (q.y.c2)
        let qxmax: float = max3 (q.x.c0) (q.x.c1) (q.x.c2)
        let qymax: float = max3 (q.y.c0) (q.y.c1) (q.y.c2)
        let mutable exclude: bool = true
        let mutable accept: bool = false
        let mutable inter: Point = { x = 0.0; y = 0.0 }
        if unbox<bool> (rectsOverlap pxmin pymin pxmax pymax qxmin qymin qxmax qymax) then
            exclude <- false
            let xmin: float = maxf pxmin qxmin
            let xmax: float = minf pxmax qxmax
            if (xmax - xmin) <= tol then
                let ymin: float = maxf pymin qymin
                let ymax: float = minf pymax qymax
                if (ymax - ymin) <= tol then
                    accept <- true
                    inter <- { inter with x = 0.5 * (xmin + xmax) }
                    inter <- { inter with y = 0.5 * (ymin + ymax) }
        __ret <- Map.ofList [("exclude", box exclude); ("accept", box accept); ("intersect", box inter)]
        raise Return
        __ret
    with
        | Return -> __ret
and seemsToBeDuplicate (pts: Point array) (xy: Point) (spacing: float) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable pts = pts
    let mutable xy = xy
    let mutable spacing = spacing
    try
        let mutable i: int = 0
        while i < (int (Array.length pts)) do
            let pt: Point = pts.[i]
            if ((float (absf ((pt.x) - (xy.x)))) < spacing) && ((float (absf ((pt.y) - (xy.y)))) < spacing) then
                __ret <- true
                raise Return
            i <- i + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and findIntersects (p: QuadCurve) (q: QuadCurve) (tol: float) (spacing: float) =
    let mutable __ret : Point array = Unchecked.defaultof<Point array>
    let mutable p = p
    let mutable q = q
    let mutable tol = tol
    let mutable spacing = spacing
    try
        let mutable inters: Point array = [||]
        let mutable workload: Map<string, QuadCurve> array = [|{ p = p; q = q }|]
        while (int (Array.length workload)) > 0 do
            let idx: int = (int (Array.length workload)) - 1
            let work: Map<string, QuadCurve> = workload.[idx]
            workload <- Array.sub workload 0 (idx - 0)
            let res: Map<string, obj> = testIntersect (unbox<QuadCurve> (work.["p"] |> unbox<QuadCurve>)) (unbox<QuadCurve> (work.["q"] |> unbox<QuadCurve>)) tol
            let excl: obj = box (res.["exclude"])
            let acc: obj = box (res.["accept"])
            let inter: Point = unbox<Point> (res.["intersect"])
            if unbox<bool> acc then
                if not (seemsToBeDuplicate inters inter spacing) then
                    inters <- Array.append inters [|inter|]
            else
                if not excl then
                    let ps: QuadCurve array = subdivideQuadCurve (unbox<QuadCurve> (work.["p"] |> unbox<QuadCurve>)) 0.5
                    let qs: QuadCurve array = subdivideQuadCurve (unbox<QuadCurve> (work.["q"] |> unbox<QuadCurve>)) 0.5
                    let p0: QuadCurve = ps.[0]
                    let p1: QuadCurve = ps.[1]
                    let q0: QuadCurve = qs.[0]
                    let q1: QuadCurve = qs.[1]
                    workload <- Array.append workload [|Map.ofList [("p", p0); ("q", q0)]|]
                    workload <- Array.append workload [|Map.ofList [("p", p0); ("q", q1)]|]
                    workload <- Array.append workload [|Map.ofList [("p", p1); ("q", q0)]|]
                    workload <- Array.append workload [|Map.ofList [("p", p1); ("q", q1)]|]
        __ret <- inters
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let p: QuadCurve = { x = { c0 = -1.0; c1 = 0.0; c2 = 1.0 }; y = { c0 = 0.0; c1 = 10.0; c2 = 0.0 } }
        let q: QuadCurve = { x = { c0 = 2.0; c1 = -8.0; c2 = 2.0 }; y = { c0 = 1.0; c1 = 2.0; c2 = 3.0 } }
        let tol: float = 0.0000001
        let spacing: float = tol * 10.0
        let inters: Point array = findIntersects p q tol spacing
        let mutable i: int = 0
        while i < (int (Array.length inters)) do
            let pt: Point = inters.[i]
            printfn "%s" (((("(" + (string (pt.x))) + ", ") + (string (pt.y))) + ")")
            i <- i + 1
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()

File: tests/rosetta/transpiler/Pascal/b-zier-curves-intersections.pas
{$mode objfpc}
program Main;
uses SysUtils, fgl;
type PointArray = array of Point;
type QuadCurveArray = array of QuadCurve;
type QuadSplineArray = array of QuadSpline;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
type Point = record
  x: real;
  y: real;
end;
type QuadSpline = record
  c0: real;
  c1: real;
  c2: real;
end;
type QuadCurve = record
  x: QuadSpline;
  y: QuadSpline;
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
function Map6(findIntersects_p1: QuadCurve; findIntersects_q1: QuadCurve): specialize TFPGMap<string, Variant>; forward;
function Map5(findIntersects_p1: QuadCurve; findIntersects_q0: QuadCurve): specialize TFPGMap<string, Variant>; forward;
function Map4(findIntersects_p0: QuadCurve; findIntersects_q1: QuadCurve): specialize TFPGMap<string, Variant>; forward;
function Map3(findIntersects_p0: QuadCurve; findIntersects_q0: QuadCurve): specialize TFPGMap<string, Variant>; forward;
function Map2(p: QuadCurve; q: QuadCurve): specialize TFPGMap<string, Variant>; forward;
function Map1(testIntersect_accept: boolean; testIntersect_exclude: boolean; testIntersect_inter: Point): specialize TFPGMap<string, Variant>; forward;
function makeQuadCurve(x: QuadSpline; y: QuadSpline): QuadCurve; forward;
function makeQuadSpline(c0: real; c1: real; c2: real): QuadSpline; forward;
function makePoint(x: real; y: real): Point; forward;
function absf(x: real): real; forward;
function maxf(a: real; b: real): real; forward;
function minf(a: real; b: real): real; forward;
function max3(a: real; b: real; c: real): real; forward;
function min3(a: real; b: real; c: real): real; forward;
function subdivideQuadSpline(q: QuadSpline; t: real): QuadSplineArray; forward;
function subdivideQuadCurve(q: QuadCurve; t: real): QuadCurveArray; forward;
function rectsOverlap(xa0: real; ya0: real; xa1: real; ya1: real; xb0: real; yb0: real; xb1: real; yb1: real): boolean; forward;
function testIntersect(p: QuadCurve; q: QuadCurve; tol: real): specialize TFPGMap<string, Variant>; forward;
function seemsToBeDuplicate(pts: PointArray; xy: Point; spacing: real): boolean; forward;
function findIntersects(p: QuadCurve; q: QuadCurve; tol: real; spacing: real): PointArray; forward;
procedure main(); forward;
function Map6(findIntersects_p1: QuadCurve; findIntersects_q1: QuadCurve): specialize TFPGMap<string, Variant>;
begin
  Result := specialize TFPGMap<string, Variant>.Create();
  Result.AddOrSetData('p', Variant(findIntersects_p1));
  Result.AddOrSetData('q', Variant(findIntersects_q1));
end;
function Map5(findIntersects_p1: QuadCurve; findIntersects_q0: QuadCurve): specialize TFPGMap<string, Variant>;
begin
  Result := specialize TFPGMap<string, Variant>.Create();
  Result.AddOrSetData('p', Variant(findIntersects_p1));
  Result.AddOrSetData('q', Variant(findIntersects_q0));
end;
function Map4(findIntersects_p0: QuadCurve; findIntersects_q1: QuadCurve): specialize TFPGMap<string, Variant>;
begin
  Result := specialize TFPGMap<string, Variant>.Create();
  Result.AddOrSetData('p', Variant(findIntersects_p0));
  Result.AddOrSetData('q', Variant(findIntersects_q1));
end;
function Map3(findIntersects_p0: QuadCurve; findIntersects_q0: QuadCurve): specialize TFPGMap<string, Variant>;
begin
  Result := specialize TFPGMap<string, Variant>.Create();
  Result.AddOrSetData('p', Variant(findIntersects_p0));
  Result.AddOrSetData('q', Variant(findIntersects_q0));
end;
function Map2(p: QuadCurve; q: QuadCurve): specialize TFPGMap<string, Variant>;
begin
  Result := specialize TFPGMap<string, Variant>.Create();
  Result.AddOrSetData('p', Variant(p));
  Result.AddOrSetData('q', Variant(q));
end;
function Map1(testIntersect_accept: boolean; testIntersect_exclude: boolean; testIntersect_inter: Point): specialize TFPGMap<string, Variant>;
begin
  Result := specialize TFPGMap<string, Variant>.Create();
  Result.AddOrSetData('exclude', Variant(testIntersect_exclude));
  Result.AddOrSetData('accept', Variant(testIntersect_accept));
  Result.AddOrSetData('intersect', Variant(testIntersect_inter));
end;
function makeQuadCurve(x: QuadSpline; y: QuadSpline): QuadCurve;
begin
  Result.x := x;
  Result.y := y;
end;
function makeQuadSpline(c0: real; c1: real; c2: real): QuadSpline;
begin
  Result.c0 := c0;
  Result.c1 := c1;
  Result.c2 := c2;
end;
function makePoint(x: real; y: real): Point;
begin
  Result.x := x;
  Result.y := y;
end;
function absf(x: real): real;
begin
  if x < 0 then begin
  exit(-x);
end;
  exit(x);
end;
function maxf(a: real; b: real): real;
begin
  if a > b then begin
  exit(a);
end;
  exit(b);
end;
function minf(a: real; b: real): real;
begin
  if a < b then begin
  exit(a);
end;
  exit(b);
end;
function max3(a: real; b: real; c: real): real;
var
  max3_m: real;
begin
  max3_m := a;
  if b > max3_m then begin
  max3_m := b;
end;
  if c > max3_m then begin
  max3_m := c;
end;
  exit(max3_m);
end;
function min3(a: real; b: real; c: real): real;
var
  min3_m: real;
begin
  min3_m := a;
  if b < min3_m then begin
  min3_m := b;
end;
  if c < min3_m then begin
  min3_m := c;
end;
  exit(min3_m);
end;
function subdivideQuadSpline(q: QuadSpline; t: real): QuadSplineArray;
var
  subdivideQuadSpline_s: real;
  subdivideQuadSpline_u: QuadSpline;
  subdivideQuadSpline_v: QuadSpline;
begin
  subdivideQuadSpline_s := 1 - t;
  subdivideQuadSpline_u := makeQuadSpline(q.c0, 0, 0);
  subdivideQuadSpline_v := makeQuadSpline(0, 0, q.c2);
  subdivideQuadSpline_u.c1 := (subdivideQuadSpline_s * q.c0) + (t * q.c1);
  subdivideQuadSpline_v.c1 := (subdivideQuadSpline_s * q.c1) + (t * q.c2);
  subdivideQuadSpline_u.c2 := (subdivideQuadSpline_s * subdivideQuadSpline_u.c1) + (t * subdivideQuadSpline_v.c1);
  subdivideQuadSpline_v.c0 := subdivideQuadSpline_u.c2;
  exit([subdivideQuadSpline_u, subdivideQuadSpline_v]);
end;
function subdivideQuadCurve(q: QuadCurve; t: real): QuadCurveArray;
var
  subdivideQuadCurve_xs: QuadSplineArray;
  subdivideQuadCurve_ys: QuadSplineArray;
  subdivideQuadCurve_u: QuadCurve;
  subdivideQuadCurve_v: QuadCurve;
begin
  subdivideQuadCurve_xs := subdivideQuadSpline(q.x, t);
  subdivideQuadCurve_ys := subdivideQuadSpline(q.y, t);
  subdivideQuadCurve_u := makeQuadCurve(subdivideQuadCurve_xs[0], subdivideQuadCurve_ys[0]);
  subdivideQuadCurve_v := makeQuadCurve(subdivideQuadCurve_xs[1], subdivideQuadCurve_ys[1]);
  exit([subdivideQuadCurve_u, subdivideQuadCurve_v]);
end;
function rectsOverlap(xa0: real; ya0: real; xa1: real; ya1: real; xb0: real; yb0: real; xb1: real; yb1: real): boolean;
begin
  exit((((xb0 <= xa1) and (xa0 <= xb1)) and (yb0 <= ya1)) and (ya0 <= yb1));
end;
function testIntersect(p: QuadCurve; q: QuadCurve; tol: real): specialize TFPGMap<string, Variant>;
var
  testIntersect_pxmin: real;
  testIntersect_pymin: real;
  testIntersect_pxmax: real;
  testIntersect_pymax: real;
  testIntersect_qxmin: real;
  testIntersect_qymin: real;
  testIntersect_qxmax: real;
  testIntersect_qymax: real;
  testIntersect_exclude: boolean;
  testIntersect_accept: boolean;
  testIntersect_inter: Point;
  testIntersect_xmin: real;
  testIntersect_xmax: real;
  testIntersect_ymin: real;
  testIntersect_ymax: real;
begin
  testIntersect_pxmin := min3(p.x.c0, p.x.c1, p.x.c2);
  testIntersect_pymin := min3(p.y.c0, p.y.c1, p.y.c2);
  testIntersect_pxmax := max3(p.x.c0, p.x.c1, p.x.c2);
  testIntersect_pymax := max3(p.y.c0, p.y.c1, p.y.c2);
  testIntersect_qxmin := min3(q.x.c0, q.x.c1, q.x.c2);
  testIntersect_qymin := min3(q.y.c0, q.y.c1, q.y.c2);
  testIntersect_qxmax := max3(q.x.c0, q.x.c1, q.x.c2);
  testIntersect_qymax := max3(q.y.c0, q.y.c1, q.y.c2);
  testIntersect_exclude := true;
  testIntersect_accept := false;
  testIntersect_inter := makePoint(0, 0);
  if rectsOverlap(testIntersect_pxmin, testIntersect_pymin, testIntersect_pxmax, testIntersect_pymax, testIntersect_qxmin, testIntersect_qymin, testIntersect_qxmax, testIntersect_qymax) then begin
  testIntersect_exclude := false;
  testIntersect_xmin := maxf(testIntersect_pxmin, testIntersect_qxmin);
  testIntersect_xmax := minf(testIntersect_pxmax, testIntersect_qxmax);
  if (testIntersect_xmax - testIntersect_xmin) <= tol then begin
  testIntersect_ymin := maxf(testIntersect_pymin, testIntersect_qymin);
  testIntersect_ymax := minf(testIntersect_pymax, testIntersect_qymax);
  if (testIntersect_ymax - testIntersect_ymin) <= tol then begin
  testIntersect_accept := true;
  testIntersect_inter.x := 0.5 * (testIntersect_xmin + testIntersect_xmax);
  testIntersect_inter.y := 0.5 * (testIntersect_ymin + testIntersect_ymax);
end;
end;
end;
  exit(Map1(testIntersect_accept, testIntersect_exclude, testIntersect_inter));
end;
function seemsToBeDuplicate(pts: PointArray; xy: Point; spacing: real): boolean;
var
  seemsToBeDuplicate_i: integer;
  seemsToBeDuplicate_pt: Point;
begin
  seemsToBeDuplicate_i := 0;
  while seemsToBeDuplicate_i < Length(pts) do begin
  seemsToBeDuplicate_pt := pts[seemsToBeDuplicate_i];
  if (absf(seemsToBeDuplicate_pt.x - xy.x) < spacing) and (absf(seemsToBeDuplicate_pt.y - xy.y) < spacing) then begin
  exit(true);
end;
  seemsToBeDuplicate_i := seemsToBeDuplicate_i + 1;
end;
  exit(false);
end;
function findIntersects(p: QuadCurve; q: QuadCurve; tol: real; spacing: real): PointArray;
var
  findIntersects_inters: array of Point;
  findIntersects_workload: array of specialize TFPGMap<string, QuadCurve>;
  findIntersects_idx: integer;
  findIntersects_work: specialize TFPGMap<string, QuadCurve>;
  findIntersects_res: specialize TFPGMap<string, Variant>;
  findIntersects_excl: integer;
  findIntersects_excl_idx: integer;
  findIntersects_acc: integer;
  findIntersects_acc_idx: integer;
  findIntersects_inter: integer;
  findIntersects_inter_idx: integer;
  findIntersects_ps: QuadCurveArray;
  findIntersects_qs: QuadCurveArray;
  findIntersects_p0: QuadCurve;
  findIntersects_p1: QuadCurve;
  findIntersects_q0: QuadCurve;
  findIntersects_q1: QuadCurve;
begin
  findIntersects_inters := [];
  findIntersects_workload := [Map2(p, q)];
  while Length(findIntersects_workload) > 0 do begin
  findIntersects_idx := Length(findIntersects_workload) - 1;
  findIntersects_work := findIntersects_workload[findIntersects_idx];
  findIntersects_workload := copy(findIntersects_workload, 0, findIntersects_idx);
  findIntersects_res := testIntersect(findIntersects_work['p'], findIntersects_work['q'], tol);
  findIntersects_excl_idx := findIntersects_res.IndexOf('exclude');
  if findIntersects_excl_idx <> -1 then begin
  findIntersects_excl := findIntersects_res.Data[findIntersects_excl_idx];
end else begin
  findIntersects_excl := 0;
end;
  findIntersects_acc_idx := findIntersects_res.IndexOf('accept');
  if findIntersects_acc_idx <> -1 then begin
  findIntersects_acc := findIntersects_res.Data[findIntersects_acc_idx];
end else begin
  findIntersects_acc := 0;
end;
  findIntersects_inter_idx := findIntersects_res.IndexOf('intersect');
  if findIntersects_inter_idx <> -1 then begin
  findIntersects_inter := findIntersects_res.Data[findIntersects_inter_idx];
end else begin
  findIntersects_inter := 0;
end;
  if findIntersects_acc then begin
  if not seemsToBeDuplicate(findIntersects_inters, findIntersects_inter, spacing) then begin
  findIntersects_inters := concat(findIntersects_inters, [findIntersects_inter]);
end;
end else begin
  if not findIntersects_excl then begin
  findIntersects_ps := subdivideQuadCurve(findIntersects_work['p'], 0.5);
  findIntersects_qs := subdivideQuadCurve(findIntersects_work['q'], 0.5);
  findIntersects_p0 := findIntersects_ps[0];
  findIntersects_p1 := findIntersects_ps[1];
  findIntersects_q0 := findIntersects_qs[0];
  findIntersects_q1 := findIntersects_qs[1];
  findIntersects_workload := concat(findIntersects_workload, [Map3(findIntersects_p0, findIntersects_q0)]);
  findIntersects_workload := concat(findIntersects_workload, [Map4(findIntersects_p0, findIntersects_q1)]);
  findIntersects_workload := concat(findIntersects_workload, [Map5(findIntersects_p1, findIntersects_q0)]);
  findIntersects_workload := concat(findIntersects_workload, [Map6(findIntersects_p1, findIntersects_q1)]);
end;
end;
end;
  exit(findIntersects_inters);
end;
procedure main();
var
  main_p: QuadCurve;
  main_q: QuadCurve;
  main_tol: real;
  main_spacing: real;
  main_inters: PointArray;
  main_i: integer;
  main_pt: Point;
begin
  main_p := makeQuadCurve((c0: -1; c1: 0; c2: 1), (c0: 0; c1: 10; c2: 0));
  main_q := makeQuadCurve((c0: 2; c1: -8; c2: 2), (c0: 1; c1: 2; c2: 3));
  main_tol := 1e-07;
  main_spacing := main_tol * 10;
  main_inters := findIntersects(main_p, main_q, main_tol, main_spacing);
  main_i := 0;
  while main_i < Length(main_inters) do begin
  main_pt := main_inters[main_i];
  writeln(((('(' + FloatToStr(main_pt.x)) + ', ') + FloatToStr(main_pt.y)) + ')');
  main_i := main_i + 1;
end;
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  main();
  Sleep(1);
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
